File: src/org.eclipse.ice.reactor/src/org/eclipse/ice/reactor/LWRComponentReader.java
/*******************************************************************************
 * Copyright (c) 2013, 2014 UT-Battelle, LLC.
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License v1.0
 * which accompanies this distribution, and is available at
 * http://www.eclipse.org/legal/epl-v10.html
 *
 * Contributors:
 *   Initial API and implementation and/or initial documentation - Jay Jay Billings,
 *   Jordan H. Deyton, Dasha Gorin, Alexander J. McCaskey, Taylor Patterson,
 *   Claire Saunders, Matthew Wang, Anna Wojtowicz
 *******************************************************************************/
package org.eclipse.ice.reactor;

import java.net.URI;
import java.util.ArrayList;
import java.util.HashMap;

import org.eclipse.ice.io.hdf.HdfFileFactory;
import org.eclipse.ice.io.hdf.HdfReaderFactory;
import org.eclipse.ice.io.hdf.IHdfReadable;
import org.eclipse.ice.io.hdf.IHdfReader;
import org.eclipse.ice.reactor.bwr.BWReactor;
import org.eclipse.ice.reactor.pwr.*;

import ncsa.hdf.object.h5.H5File;
import ncsa.hdf.object.h5.H5Group;

/**
 * <p>
 * The LWRComponentReader class creates and populates an LWRComponent or
 * LWRComposite tree from an HDF5 file. This class implements the IHdfReadable
 * interface. This takes any type of LWRComponent or LWRComposite and iterates
 * from the top of that particular tree down, so a user could read from HDF5 and
 * populate that list accordingly to any part of a Reactor or its delegated
 * classes that inherit from LWRComponent.
 * </p>
 * 
 * @author s4h
 */
public class LWRComponentReader implements IHdfReader {

	/**
	 * <p>
	 * A HashMap keyed on HDF5LWRTagType name storing unique LWRComponent
	 * instances.
	 * </p>
	 */
	private HashMap<HDF5LWRTagType, LWRComponent> lWRComponentInstanceMap;

	/**
	 * The Constructor.
	 */
	public LWRComponentReader() {

		// Create a new HashMap<HDF5LWRTagType, LWRComponent> for the
		// LWRComponent instances
		this.lWRComponentInstanceMap = new HashMap<HDF5LWRTagType, LWRComponent>();

		// Create and assign default instances to the map
		this.lWRComponentInstanceMap.put(HDF5LWRTagType.CONTROL_BANK,
				new ControlBank());
		this.lWRComponentInstanceMap.put(HDF5LWRTagType.FUEL_ASSEMBLY,
				new FuelAssembly(-1));
		this.lWRComponentInstanceMap.put(HDF5LWRTagType.GRID_LABEL_PROVIDER,
				new GridLabelProvider(-1));
		this.lWRComponentInstanceMap.put(HDF5LWRTagType.INCORE_INSTRUMENT,
				new IncoreInstrument());
		this.lWRComponentInstanceMap.put(HDF5LWRTagType.LWRROD, new LWRRod());
		this.lWRComponentInstanceMap.put(HDF5LWRTagType.MATERIAL,
				new Material());
		this.lWRComponentInstanceMap.put(HDF5LWRTagType.PWREACTOR,
				new PressurizedWaterReactor(-1));
		this.lWRComponentInstanceMap.put(HDF5LWRTagType.RING, new Ring());
		this.lWRComponentInstanceMap.put(HDF5LWRTagType.ROD_CLUSTER_ASSEMBLY,
				new RodClusterAssembly(-1));
		this.lWRComponentInstanceMap.put(HDF5LWRTagType.MATERIALBLOCK,
				new MaterialBlock());
		this.lWRComponentInstanceMap.put(HDF5LWRTagType.TUBE, new Tube());
		this.lWRComponentInstanceMap.put(HDF5LWRTagType.LWRCOMPONENT,
				new LWRComponent());
		this.lWRComponentInstanceMap.put(HDF5LWRTagType.LWRCOMPOSITE,
				new LWRComposite());
		this.lWRComponentInstanceMap.put(HDF5LWRTagType.LWREACTOR,
				new LWReactor(-1));
		this.lWRComponentInstanceMap.put(HDF5LWRTagType.BWREACTOR,
				new BWReactor(-1));
		this.lWRComponentInstanceMap.put(HDF5LWRTagType.PWRASSEMBLY,
				new PWRAssembly(-1));
		this.lWRComponentInstanceMap.put(HDF5LWRTagType.LWRGRIDMANAGER,
				new LWRGridManager(-1));

	}

	/**
	 * (non-Javadoc)
	 * 
	 * @see IHdfReader#read(URI uri)
	 */
	public IHdfReadable read(URI uri) {

		// Open the file at the provided uri
		H5File h5File = HdfFileFactory.openH5File(uri);

		// If the file is null, then return false
		if (h5File == null) {
			return null;
		}
		// Get the root group from the file
		H5Group rootH5Group = (H5Group) ((javax.swing.tree.DefaultMutableTreeNode) h5File
				.getRootNode()).getUserObject();

		// If the root group is null, then return null
		if (rootH5Group == null) {
			return null;
		}
		// Get the first child group from the root group
		H5Group h5Group = HdfReaderFactory.getChildH5Group(rootH5Group, 0);

		// Read the group into the iHdfReadable
		IHdfReadable iHdfReadable = this.read(h5Group);

		// Close the file
		HdfFileFactory.closeH5File(h5File);

		// Return the iHdfReadable
		return iHdfReadable;

	}

	/**
	 * <p>
	 * Returns a clone of the LWRComponent instance corresponding to the
	 * provided HDF5LWRTagType enumeration literal.
	 * </p>
	 * 
	 * @param HDF5LWRTag
	 *            <p>
	 *            A HDF5LWRTagType enumeration literal.
	 *            </p>
	 * @return <p>
	 *         A clone of the LWRComponent instance corresponding to the
	 *         provided HDF5LWRTagType enumeration literal.
	 *         </p>
	 */
	private LWRComponent getLWRComponentInstance(HDF5LWRTagType HDF5LWRTag) {

		// Get the LWRComponent instance corresponding to the provided tag
		LWRComponent lWRComponent = (LWRComponent) lWRComponentInstanceMap.get(
				HDF5LWRTag).clone();

		// Return the instance
		return lWRComponent;
	}

	/**
	 * <p>
	 * Recursively reads an H5Group and its child H5Groups and Datasets into an
	 * IHdfReadable.
	 * </p>
	 * 
	 * @param h5Group
	 *            <p>
	 *            The H5Group to be read.
	 *            </p>
	 * @return <p>
	 *         An IHdfReadable that has been populated by the provided H5Group
	 *         and its child H5Groups and Datasets.
	 *         </p>
	 */
	private IHdfReadable read(H5Group h5Group) {

		IHdfReadable iHdfReadable = null;

		// Check for a null h5Group
		if (h5Group == null) {
			return null;
		}
		// Get the tag from the group
		HDF5LWRTagType HDF5LWRTag = HDF5LWRTagType.toType(HdfReaderFactory
				.readStringAttribute(h5Group, "HDF5LWRTag"));

		// If the tag does not exist, skip it
		if (HDF5LWRTag != null) {

			// Get an instance from the tag
			iHdfReadable = this.getLWRComponentInstance(HDF5LWRTag);

			// Read in the attributes
			iHdfReadable.readAttributes(h5Group);

			// Read in the datasets
			iHdfReadable.readDatasets(h5Group);

			// Get a list of child groups
			ArrayList<H5Group> childGroupList = HdfReaderFactory
					.getChildH5Groups(h5Group);

			// If the group has entries
			if (!(childGroupList.isEmpty())) {

				// Cycle over these entries
				for (H5Group group : childGroupList) {

					// Create and populate a child from the child group
					IHdfReadable child = this.read(group);

					// Read the child into the readable
					iHdfReadable.readChild(child);
				}

			}
		}
		// Return the top level readable
		return iHdfReadable;

	}
}

File: native/org.eclipse.ice.reactor.native/LWRComponentReader.cpp
/*******************************************************************************
 * Copyright (c) 2012, 2014 UT-Battelle, LLC.
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License v1.0
 * which accompanies this distribution, and is available at
 * http://www.eclipse.org/legal/epl-v10.html
 *
 * Contributors:
 *   Initial API and implementation and/or initial documentation - Jay Jay Billings,
 *   Jordan H. Deyton, Dasha Gorin, Alexander J. McCaskey, Taylor Patterson,
 *   Claire Saunders, Matthew Wang, Anna Wojtowicz
 *******************************************************************************/

#include "LWRComponentReader.h"
#include "UtilityOperations.h"
#include <HdfReaderFactory.h>
#include <HdfFileFactory.h>
#include <map>
#include "pwr/ControlBank.h"
#include "pwr/FuelAssembly.h"
#include "pwr/IncoreInstrument.h"
#include "pwr/PWRAssembly.h"
#include "pwr/PressurizedWaterReactor.h"
#include "pwr/RodClusterAssembly.h"
#include "bwr/BWReactor.h"
#include "GridLabelProvider.h"
#include "LWRRod.h"
#include "Material.h"
#include "MaterialBlock.h"
#include "Tube.h"
#include "LWRComponent.h"
#include "LWRComposite.h"
#include "LWReactor.h"
#include "LWRGridManager.h"
#include <H5Exception.h>

using namespace ICE_Reactor;

/**
 * Copy constructor.  Not used
 */
LWRComponentReader::LWRComponentReader(LWRComponentReader & arg) {


}

/**
 * Destructor
 */
LWRComponentReader::~LWRComponentReader() {
    //TODO Auto-generated method stub
}

/**
 * Constructor
 */
LWRComponentReader::LWRComponentReader() {
    // begin-user-code

    //Create and assign default instances to the map
    this->lWRComponentInstanceMap.insert( std::pair<HDF5LWRTagType, std::shared_ptr <LWRComponent> >( CONTROL_BANK, std::shared_ptr<LWRComponent> ( new ControlBank())));
    this->lWRComponentInstanceMap.insert( std::pair<HDF5LWRTagType, std::shared_ptr <LWRComponent> >( FUEL_ASSEMBLY, std::shared_ptr<LWRComponent> ( new FuelAssembly(-1))));
    this->lWRComponentInstanceMap.insert( std::pair<HDF5LWRTagType, std::shared_ptr <LWRComponent> >( GRID_LABEL_PROVIDER, std::shared_ptr<LWRComponent> ( new GridLabelProvider(-1))));
    this->lWRComponentInstanceMap.insert( std::pair<HDF5LWRTagType, std::shared_ptr <LWRComponent> >( INCORE_INSTRUMENT, std::shared_ptr<LWRComponent> ( new IncoreInstrument())));
    this->lWRComponentInstanceMap.insert( std::pair<HDF5LWRTagType, std::shared_ptr <LWRComponent> >( LWRROD, std::shared_ptr<LWRComponent> ( new LWRRod())));
    this->lWRComponentInstanceMap.insert( std::pair<HDF5LWRTagType, std::shared_ptr <LWRComponent> >( MATERIAL, std::shared_ptr<LWRComponent> ( new Material())));
    this->lWRComponentInstanceMap.insert( std::pair<HDF5LWRTagType, std::shared_ptr <LWRComponent> >( PWREACTOR, std::shared_ptr<LWRComponent> ( new PressurizedWaterReactor(-1))));
    this->lWRComponentInstanceMap.insert( std::pair<HDF5LWRTagType, std::shared_ptr <LWRComponent> >( RING, std::shared_ptr<LWRComponent> ( new Ring())));
    this->lWRComponentInstanceMap.insert( std::pair<HDF5LWRTagType, std::shared_ptr <LWRComponent> >( ROD_CLUSTER_ASSEMBLY, std::shared_ptr<LWRComponent> ( new RodClusterAssembly(-1))));
    this->lWRComponentInstanceMap.insert( std::pair<HDF5LWRTagType, std::shared_ptr <LWRComponent> >( MATERIALBLOCK, std::shared_ptr<LWRComponent> ( new MaterialBlock())));
    this->lWRComponentInstanceMap.insert( std::pair<HDF5LWRTagType, std::shared_ptr <LWRComponent> >( TUBE, std::shared_ptr<LWRComponent> ( new Tube())));
    this->lWRComponentInstanceMap.insert( std::pair<HDF5LWRTagType, std::shared_ptr <LWRComponent> >( LWRCOMPONENT, std::shared_ptr<LWRComponent> ( new LWRComponent())));
    this->lWRComponentInstanceMap.insert( std::pair<HDF5LWRTagType, std::shared_ptr <LWRComponent> >( LWRCOMPOSITE, std::shared_ptr<LWRComponent> ( new LWRComposite())));
    this->lWRComponentInstanceMap.insert( std::pair<HDF5LWRTagType, std::shared_ptr <LWRComponent> >( LWREACTOR, std::shared_ptr<LWRComponent> ( new LWReactor(-1))));
    this->lWRComponentInstanceMap.insert( std::pair<HDF5LWRTagType, std::shared_ptr <LWRComponent> >( BWREACTOR, std::shared_ptr<LWRComponent> ( new BWReactor(-1))));
    this->lWRComponentInstanceMap.insert( std::pair<HDF5LWRTagType, std::shared_ptr <LWRComponent> >( PWRASSEMBLY, std::shared_ptr<LWRComponent> ( new PWRAssembly(-1))));
    this->lWRComponentInstanceMap.insert( std::pair<HDF5LWRTagType, std::shared_ptr <LWRComponent> >( LWRGRIDMANAGER, std::shared_ptr<LWRComponent> ( new LWRGridManager(-1))));

    // end-user-code

}

/**
 * Parses a file given a path from HDF5 to a IHdfReadable object.
 *
 * @param the path
 *
 * @return the IHDF5Readalbe.
 */
std::shared_ptr<ICE_IO::IHdfReadable> LWRComponentReader::read(const std::string path) {
    // begin-user-code

    //Local Declarations
    std::shared_ptr<ICE_IO::IHdfReadable> nullReadable;
    std::shared_ptr<ICE_IO::IHdfReadable> iHdfReadable;

    try {
        //Open the file at the provided uri
        std::shared_ptr<H5::H5File> h5File = ICE_IO::HdfFileFactory::openH5File(path);

        //If the file is null, then return false
        if (h5File.get()== NULL) return nullReadable;


        //Get the root group from the file
        std::shared_ptr<H5::Group> rootH5Group ( new H5::Group( h5File.get()->openGroup("/")));

        //If the root group is null, then return null
        if (rootH5Group.get() == NULL) return nullReadable;


        //Get the first child group from the root group
        std::shared_ptr<H5::Group> h5Group = ICE_IO::HdfReaderFactory::getChildH5Group(rootH5Group, 0);

        //Read the group into the iHdfReadable
        iHdfReadable = LWRComponentReader::read(h5Group);

        //Close the file
        ICE_IO::HdfFileFactory::closeH5File(h5File);

    } catch (...) {
        //If error caught, return null readable
        return nullReadable;
    }

    //Return the iHdfReadable
    return iHdfReadable;

    // end-user-code

}

/**
	 * Returns a new instance of the HDF5LWRTagType or null
	 *
	 * @param  The tag
	 *
	 * @return LWRComponent
	 */
std::shared_ptr<LWRComponent> LWRComponentReader::getLWRComponentInstance(HDF5LWRTagType HDF5LWRTag) {

    //Get the LWRComponent instance corresponding to the provided tag
    std::shared_ptr<ICE_DS::Identifiable> identifiable (LWRComponentReader::lWRComponentInstanceMap.at(HDF5LWRTag).get()->clone());

    //Cast as LWRComponent
    std::shared_ptr<LWRComponent> lWRComponent = std::dynamic_pointer_cast <LWRComponent> (identifiable);

    //Return the instance
    return lWRComponent;
}

/**
 * Reads the H5Group and returns the list of IHDF5Readables
 *
 * @param The Group
 *
 * @return the IHdfReadable
 */
std::shared_ptr<ICE_IO::IHdfReadable> LWRComponentReader::read(std::shared_ptr<H5::Group> h5Group) {

    //begin-user-code

    //Local Declarations
    std::shared_ptr<ICE_IO::IHdfReadable> iHdfReadable;
    std::shared_ptr<ICE_IO::IHdfReadable> nullReadable;
    HDF5LWRTagType HDF5LWRTag;

    //Check for a null h5Group
    if (h5Group.get() == NULL) return nullReadable;


    try {
        //Turn off errors for finding the tag.  If the tag does not exist, then in this try block do not worry about it.
        H5::Exception::dontPrint();
        //Get the tag from the group
        HDF5LWRTag = UtilityOperations::fromStringHDF5Tag(ICE_IO::HdfReaderFactory::readStringAttribute(h5Group, "HDF5LWRTag"));
    } catch (...) {
        //Tag does not exist, then return
        return nullReadable;
    }

    try {
        //If the tag does not exist, skip it
        if(HDF5LWRTag) {
            //Get an instance from the tag
            iHdfReadable = LWRComponentReader::getLWRComponentInstance(HDF5LWRTag);

            //Read in the attributes
            iHdfReadable.get()->readAttributes(h5Group);

            //Read in the datasets
            iHdfReadable.get()->readDatasets(h5Group);

            //Get a list of child groups
            std::vector<std::shared_ptr<H5::Group> > childGroupList = ICE_IO::HdfReaderFactory::getChildH5Groups(h5Group);

            //If the group has entries
            if (childGroupList.size() > 0) {

                //Cycle over these entries
                for (int i = 0; i < childGroupList.size(); i++ ) {

                    //Create and populate a child from the child group
                    std::shared_ptr<ICE_IO::IHdfReadable> child = LWRComponentReader::read(childGroupList.at(i));

                    //Read the child into the readable
                    if(child.get() != NULL) {
                        iHdfReadable.get()->readChild(child);
                    }


                }

            }
        }

        //Return the top level readable
        return iHdfReadable;
    } catch (...) {
        //Throw it up if error
        throw -1;
    }

    //end-user-code

}
